### ISSUE TITLE
Segmentation fault on shard 0

### ISSUE THREAD

**Issue by jeffRTC** [scylladb/seastar#944]
Couldn't find a good coroutine example but the one I come up seg fault right after printing TCP message. 

```
Segmentation fault on shard 0.
Backtrace:
  0x135d8e
  0xfd790
  0xfdf0c
  0xfdf92
  0x29cb3ee361ef
  /lib/x86_64-linux-gnu/libc.so.6+0x97768
  0xe1741
  0xbe77e
  0xb389c
  0xabb1f
  0xab4ed
  0xa33fa
  0x8868e
  0xa57e6
  0x9f6be
  0xf9887
  0xf9bb0
  0x123e55
  0x12499d
  0xd1556
  0xd1e6b
  0xd1f9b
  0x8ff02
  /lib/x86_64-linux-gnu/libc.so.6+0x28564
  0x86d4d
Segmentation fault (core dumped)
```

Coroutines feels like heaven compared to the future callback hell. Here is my code. 

```
#include <iostream>
#include <string>
#include <seastar/util/std-compat.hh>
#include <seastar/core/app-template.hh>
#include <seastar/core/coroutine.hh>
#include <seastar/core/seastar.hh>
#include <seastar/core/loop.hh>
#include <seastar/core/reactor.hh>
#include <seastar/core/future-util.hh>
#include <seastar/net/api.hh>


const char* canned_response = "HTTP/1.1 200 OK\r\nContent-Length: 13\r\nConnection: close\r\n\r\nHello, world!";

seastar::future<> handle_connection(seastar::connected_socket socket, seastar::socket_address peer) {

    std::cout << "Accepted connection from " << peer << "\n";

    auto out = socket.output();
    auto in = socket.input();

    while(true) {
        seastar::data_source::tmp_buf buffer = co_await in.read();

        if(buffer) {
            std::string request(buffer.get(), buffer.size());

            std::cout << request << "\n";

            out.write(canned_response);
            out.flush();
        } else {
            // We may have to test this by removing this part 
            // Eg TLS issues

            std::cout << "Peer disconnected from " << peer << "\n";

            break;
        }
    }

    co_await out.close();
}

seastar::future<> service_loop() {
    seastar::listen_options lo;

    // This will allow us to re run the app without port binding errors
    lo.reuse_address = true;

    auto listener = seastar::listen(seastar::make_ipv4_address({1234}), lo);

    while(true) {
        seastar::accept_result session = co_await listener.accept();

        seastar::connected_socket socket = std::move(session.connection);
        seastar::socket_address peer = session.remote_address;
        

        // This will allow us to loop without awaiting
        (void)handle_connection(std::move(socket), std::move(peer));
    }
}

int main(int argc, char** argv) {
    // Give the ability to kill the program
    seastar::app_template::config config;
    config.auto_handle_sigint_sigterm = false;

    seastar::app_template app{config};

    app.run(argc, argv, [] () -> seastar::future<> {
        co_await seastar::parallel_for_each(boost::irange<unsigned>(0, seastar::smp::count),
                [] (unsigned i) -> seastar::future<> {
            co_await seastar::smp::submit_to(i, service_loop);
        });
    });
}
```

**Comment by jeffRTC** [comment]
@avikivity
@nyh

After commenting lot of of stuff, I figured out exact line this happens,

It's this `std::string request(buffer.get(), buffer.size());`

But, I'm not exactly sure if that's the exact line because I could also print the request string. 

Anyways, I compiled with Clang and the seg fault magically disappeared. 

Is this a GCC bug?

**Comment by nyh** [comment]
I think there may be a bug in your own code:
```c++
            std::string request(buffer.get(), buffer.size());
            std::cout << request << "\n";
            out.write(canned_response);
            out.flush();
```

But you don't wait for the write() or flush() to finish.. You shouldn't call write() and flush() in parallel (and also close() below!). Although I don't know if this causes your crash. Please try to add co_await() for both write() and flush() and see if it solves the bug.

If that's your bug, then it's not a Seastar issue so please close this issue.

**Comment by jeffRTC** [comment]
@nyh 

Even with co_await, gcc compiled program crash while clang is okay..

**Comment by JasonYuchen** [comment]
@jeffRTC 
If you are using coroutine with ASAN, this is a known issue in [GCC](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95137)

**Comment by jeffRTC** [comment]
@JasonYuchen 

Oh yeah! I compile with `-fsanitize=undefined`

Thanks, I didn't know about this..

**Comment by nyh** [comment]
We definitely need some text in README.md on what the level of support of coroutines - which compilers are known to need specific parameters or have specific bugs in specific cases, with links to the bugs.

A README.md patch like this can mark Fixes this issue.

**Comment by nyh** [comment]
Actually, closing in favor of #945
